### ISSUE TITLE
AuthorizationManager should add support check like AuthenticationProvider

### ISSUE THREAD

**Issue by abccbaandy** [spring-projects/spring-security#15317]
**Expected Behavior**
Can be safely cast the `Authentication` principal to to `myPrincipal` in the `check(Supplier<Authentication> authentication, T object)` method.

**Current Behavior**
Have no idea if the `MyAuthorizationManager` can handle the `Authentication`.

**Context**
Need extra check if `Authentication` is instanceof  `MyAuthentication`.
If have a method like
````java
boolean supports(Class<?> authentication)
````
Then we can put the check code in the method, which make the authorization check code more clean.

**Comment by jzheaux** [comment]
I see your meaning, you want to be able to safely cast, and `supports` gives you the confidence that you can.

That said, you can safely cast in the following way as well:

```java
MyAuthorizationManager implement AuthorizationManager{
  check(Supplier<Authentication> authentication, T object) {
    if (!(authentication.get() instanceof MyAuthentication myAuthentication)) {
         return null;
    }
    // ...
  }
}
```

I realize that it is a little different than what you are used to in `AuthenticationProvider`. Other than the unfamiliarity, does this approach give you any trouble?

**Comment by jzheaux** [comment]
Good question. The main reason is that it seems unnecessary, and we like to keep APIs as lean as possible. Introducing this method would require that the rest of Spring Security now call `supports` in advance of calling `check`; but, if an application just needs to call `check`, then everyone gets the same outcome without the extra method.

Also, considering your interest in this, please consider following https://github.com/spring-projects/spring-security/issues/11428 which details future efforts to deprecate `AuthenticationProvider`. When you take a look at `AuthenticationManager`, `ReactiveAuthenticationManager`, `AuthorizationManager`, and `ReactiveAuthorizationManager`, you'll see that `AuthenticationProvider`, while perhaps the more familiar API for folks, is the outlier as far as API defintiion.

**Comment by jzheaux** [comment]
> Then why the AuthenticationProvider have `support` method?

I think it's just another way to conceptualize the solution. That said, the interface was introduced before my time so I don't know.

> Just like we put @ValidSomeing in the parameter

I'm not sure I follow the analogy; I think you are talking about isolating concerns, which I agree is an important design principle.

My point is that this is an intrinsic part of the authorization concern. Consider that `AuthorizationManager#check` returns `null` to mean "I abstain", which is a reasonable `AuthorizationDecision`. So even if you did have a `supports` method, you'd still have to check for `null` in the return type to see if that `AuthorizationManager` abstained.

If you have to check for `null` anyway, then it seems there is no gain from an additional method call.

(Additionally, there is the issue of the `Authentication` being deferred (through a `Supplier`), so you'd need to load it once to decide whether or not the manager supports it and again to execute the check.)

Here is the difference:

```java
if (manager.supports(authentication.get())) {
    AuthorizationDecision decision = authorizationManager.check(authentication, object);
    if (decision == null) {
        // do abstain logic
    } else if (decision.isGranted()) {
        // do granted logic
    } else {
        // do denied logic
    }
} else {
    // do abstain logic
}
```

vs

```java
AuthorizationDecision decision = authorizationManager.check(authentication, object);
if (decision == null) {
    // do abstain logic
} else if (decision.isGranted()) {
    // do granted logic
} else {
    // do denied logic
}
```

This has the benefit that the manager returning `null` means "I can't decide", which is a nice conflation of the two notions: "I don't support" and "I choose not to say whether this object is authorized". It also has the benefit that `Supplier<Authentication>#get` is called only once.
